src/osd/OSD.cc

```diff
@@ -6784,7 +6784,9 @@ void OSD::osdmap_subscribe(version_t epoch, bool force_request)
 
 void OSD::trim_maps(epoch_t oldest, int nreceived, bool skip_maps)
 {
-  epoch_t min = std::min(oldest, service.map_cache.cached_key_lower_bound());
+  epoch_t min = std::min(oldest,
+			 std::min(service.map_cache.cached_key_lower_bound(),
```

Comment:
strange enough, all osdmaps including the ones are held by PG are tracked by the `map_cache`. a `PG::osdmap_ref` is always updated when a PG moves to a new map. so `map_cache` should always return the right and a safe epoch.